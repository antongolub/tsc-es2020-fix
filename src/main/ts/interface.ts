type IFunction<A extends any[] = any[], R = any> = (...args: A) => R

export type IFixOptionsNormalized = {
  cwd: string
  debug: IFunction
  out?: string
  src?: string | string[]
  target?: string | string[]
  tsconfig: string | string[]
  dirnameVar: boolean
  filenameVar: boolean
  fillBlank?: boolean
  sourceMap?: boolean
  forceDefaultExport?: boolean
  ext: boolean | string
  unlink?: boolean
}

export type TSConfig = {
  extends?: string
  compilerOptions?: Record<string, any>
}

export type IFixOptions = Partial<Omit<IFixOptionsNormalized, 'debug'>> & {debug?: boolean | IFunction}
