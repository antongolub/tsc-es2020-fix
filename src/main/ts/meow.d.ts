// https://github.com/sindresorhus/meow/pull/246
declare module "meow" {
  import type * as meow from 'meow/build/index.d.ts'
  export = meow
}
