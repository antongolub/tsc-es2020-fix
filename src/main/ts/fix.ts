import { basename, dirname, resolve } from 'path'

import { IFixOptions, IFixOptionsNormalized } from './interface'
import {
  asArray,
  globby,
  read,
  readJson,
  resolveTsConfig,
  unixify,
  unlink,
  write,
} from './util'

export const DEFAULT_FIX_OPTIONS: IFixOptionsNormalized = {
  cwd: process.cwd(),
  tsconfig: './tsconfig.json',
  filenameVar: true,
  dirnameVar: true,
  ext: true,
  debug: false,
}

export const normalizeOptions = (
  opts?: IFixOptions,
): IFixOptionsNormalized => ({ ...DEFAULT_FIX_OPTIONS, ...opts })

export const findTargets = (
  tsconfig: string | string[],
  cwd: string,
): string[] =>
  asArray(tsconfig).reduce<string[]>((targets, file) => {
    const tsconfigJson = resolveTsConfig(resolve(cwd, file))
    const outDir = tsconfigJson?.compilerOptions?.outDir
    const module = tsconfigJson?.compilerOptions?.module?.toLowerCase()

    if (outDir && (module === 'es2020' || module === 'esnext')) {
      targets.push(outDir)
    }

    return targets
  }, [])

export const resolveDependency = (
  parent: string,
  nested: string,
  files: string[],
  cwd: string,
): string => {
  const dir = dirname(parent)
  const nmdir = resolve(cwd, 'node_modules')
  const bases = /^\..+\.[^./\\]+$/.test(nested)
    ? [nested, nested.replace(/\.[^./\\]+$/, '')]
    : [nested]
  const variants = ['.js', '.cjs', '.mjs'].reduce<string[]>((m, e) => {
    bases.forEach(v => m.push(`${v}${e}`, `${v}/index${e}`))
    return m
  }, [])

  return (
    variants.find(
      (f) =>
        files.includes(unixify(resolve(nmdir, f))) ||
        files.includes(unixify(resolve(dir, f))),
    ) || nested
  )
}

export const fixFilenameExtensions = (names: string[], ext: string): string[] =>
  names.map((name) => name.replace(/\.[^./\\]+$/, ext))

export const fixModuleReferences = (
  contents: string,
  filename: string,
  filenames: string[],
  cwd: string,
): string =>
  contents.replace(
    /(\sfrom |\simport[ (])(["'])([^"']+\/[^"']+)(["'])/g,
    (matched, control, q1, from, q2) =>
      `${control}${q1}${resolveDependency(
        filename,
        from,
        filenames,
        cwd,
      )}${q2}`,
  )

export const fixDirnameVar = (contents: string): string =>
  contents.replace(
    /__dirname/g,
    '/file:\\/\\/(.+)\\/[^/]/.exec(import.meta.url)[1]',
  ) // eslint-disable-line

export const fixFilenameVar = (contents: string): string =>
  contents.replace(/__filename/g, '/file:\\/\\/(.+)/.exec(import.meta.url)[1]') // eslint-disable-line

export const fixContents = (
  contents: string,
  filename: string,
  filenames: string[],
  { cwd, ext, dirnameVar, filenameVar }: IFixOptionsNormalized,
): string => {
  let _contents = contents

  if (ext) {
    _contents = fixModuleReferences(_contents, filename, filenames, cwd)
  }

  if (dirnameVar) {
    _contents = fixDirnameVar(_contents)
  }

  if (filenameVar) {
    _contents = fixFilenameVar(_contents)
  }

  return _contents
}

const getExtModulesWithPkgJsonExports = (cwd: string): Promise<string[]> =>
  globby(['node_modules/*/package.json'], {
    cwd: cwd,
    onlyFiles: true,
    absolute: true,
  }).then((files: string[]) =>
    files
      .filter((f: string) => readJson(f).exports)
      .map((f: string) => basename(dirname(f))),
  )

const getExtModules = async (cwd: string): Promise<string[]> =>
  globby(
    [
      'node_modules/**/*.(m|c)?js',
      '!node_modules/**/node_modules',
      ...(await getExtModulesWithPkgJsonExports(cwd)).map(
        (m: string) => `!node_modules/${m}`,
      ),
    ],
    {
      cwd: cwd,
      onlyFiles: true,
      absolute: true,
    },
  )

export const fix = async (opts?: IFixOptions): Promise<void> => {
  const _opts = normalizeOptions(opts)
  const { cwd, target, tsconfig, out = cwd, ext, debug } = _opts
  const dbg = debug ? console.log : () => {} // eslint-disable-line
  const targets = [...asArray(target), ...findTargets(tsconfig, cwd)]
  dbg('debug:cwd', cwd)
  dbg('debug:targets', targets)

  const patterns = targets.map((target) => `${target}/**/*.js`)
  const outDir = resolve(cwd, out)
  dbg('debug:outdir', outDir)

  const names = await globby(patterns, {
    cwd: cwd,
    onlyFiles: true,
    absolute: true,
  })
  const externalNames = await getExtModules(cwd)
  dbg('debug:external-names', externalNames)

  const _names =
    typeof ext === 'string' ? fixFilenameExtensions(names, ext) : names
  dbg('debug:local-names', _names)

  const allNames = [...externalNames, ..._names]
  _names.forEach((name, i) => {
    const nextName = name.replace(unixify(cwd), unixify(outDir))
    const contents = read(names[i])
    const _contents = fixContents(contents, name, allNames, _opts)

    write(nextName, _contents)

    if (cwd === outDir && nextName !== names[i]) {
      unlink(names[i])
    }
  })
}
