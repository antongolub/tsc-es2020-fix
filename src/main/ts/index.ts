export {
  fix,
  fixContents,
  fixFilenameVar,
  fixDirnameVar,
  fixRelativeModuleReferences,
  fixFileExtensions,
  DEFAULT_FIX_OPTIONS
} from './fix'
export * from './interface'
