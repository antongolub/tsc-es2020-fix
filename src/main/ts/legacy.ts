import { IFixOptionsNormalized, TFixContext } from './interface'
import {
  fixBlankFiles as _fixBlankFiles,
  fixContents as _fixContents,
  fixDefaultExport as _fixDefaultExport,
  fixDirnameVar as _fixDirnameVar,
  fixFilenameVar as _fixFilenameVar,
  fixModuleReferences as _fixModuleReferences,
  fixSourceMapRef as _fixSourceMapRef,
} from './fixes'

export const fixModuleReferences = (
  contents: string,
  filename: string,
  filenames: string[],
  cwd: string,
  ignore: string[],
): string =>
  _fixModuleReferences({contents, filename, filenames, options: {cwd}, ignore} as TFixContext).contents

export const fixDirnameVar = (contents: string, isSource?: boolean): string =>
  _fixDirnameVar({contents, isSource} as TFixContext).contents

export const fixFilenameVar = (contents: string, isSource?: boolean): string =>
  _fixFilenameVar({contents, isSource} as TFixContext).contents

export const fixDefaultExport = (contents: string): string =>
  _fixDefaultExport({contents} as TFixContext).contents

export const fixBlankFiles = (contents: string): string =>
  _fixBlankFiles({contents} as TFixContext).contents

export const fixSourceMapRef = (contents: string, originName: string, filename: string): string =>
  _fixSourceMapRef({contents, originName, filename} as TFixContext).contents

export const fixContents = (
  contents: string,
  filename: string,
  filenames: string[],
  options: IFixOptionsNormalized,
  originName = filename, // NOTE Weird contract to avoid breaking change
  isSource = false,
  ignore: string[] = [],
): string =>
  _fixContents({contents, filename, filenames, options, originName, isSource, ignore} as TFixContext).contents
