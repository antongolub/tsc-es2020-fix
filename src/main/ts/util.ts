import { readFileSync, unlinkSync } from 'fs'
import fse from 'fs-extra'
import json5 from 'json5'
import { dirname, resolve } from 'path'

import { TSConfig } from './interface'

export { globby, globbySync } from 'globby'

export const read = (file: string): string =>
  readFileSync(file, { encoding: 'utf8' })

export const write = (file: string, contents: string): void =>
  fse.outputFileSync(file, contents, { encoding: 'utf8' })

export const readJson = <D = any>(file: string): D => json5.parse(read(file))

export const asArray = <T>(value: T[] | T | undefined): T[] =>
  (value
    ? Array.isArray(value)
      ? value
      : [value]
    : [])

export const remove = unlinkSync

export const unixify = (path: string): string => path.replace(/\\/g, '/')

export const resolveTsConfig = (file: string): TSConfig => {
  const data = readJson<TSConfig>(file)

  if (data.extends) {
    const parent = resolveTsConfig(resolve(dirname(file), data.extends))

    return {
      ...parent,
      compilerOptions: { ...parent.compilerOptions, ...data.compilerOptions },
    }
  }

  return data
}
