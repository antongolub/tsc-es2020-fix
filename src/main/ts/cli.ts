#!/usr/bin/env node

import meow from 'meow'

import {fix} from './fix'

const cli = meow(`
	Usage
	  $ tsc-es2020-fix [options]

	Options
	  --tsconfig    Prod/bundle tsconfig path to search for 'outDir'
	  --target      Specify target/outDir by hand. Suppresses 'tsconfig.compilerOptions.outDir'.
	  --ext         Append extension (like '.mjs') to relative imports/re-exports
	  --dirnameVar  Replace __dirname refs with import.meta
	  --filenameVar Replace __filename with import.meta
	  --out         Output dir. Defaults to cwd, so files will be overridden
	  --cwd         cwd. process.cwd() by default

	Examples
	  $ tsc-es2020-fix --ext=.mjs --out=foo
`, {
  importMeta: import.meta,
  flags: {
    ext: {
      type: 'string'
    },
    tsconfig: {
      isMultiple: true,
      type: 'string'
    },
    dirnameVar: {
      type: 'boolean',
      default: true
    },
    filenameVar: {
      type: 'boolean',
      default: true
    },
    cwd: {
      type: 'string'
    },
    target: {
      type: 'string',
      isMultiple: true
    }
  }
})

fix(cli.flags)
