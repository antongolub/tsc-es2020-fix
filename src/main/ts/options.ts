import {IFixOptions, IFixOptionsNormalized} from "./interface";
import {omitUndefinedKeys} from './util'

export const DEFAULT_FIX_OPTIONS: IFixOptionsNormalized = {
  cwd: process.cwd(),
  tsconfig: './tsconfig.json',
  filenameVar: true,
  dirnameVar: true,
  ext: true,
  unlink: true,
  debug: () => {}, // eslint-disable-line
}

export const normalizeOptions = (
  opts?: IFixOptions,
): IFixOptionsNormalized => ({
  ...DEFAULT_FIX_OPTIONS,
  ...omitUndefinedKeys(opts || {}),
  debug: typeof opts?.debug === 'function'
    ? opts.debug
    : opts?.debug === true
      ? console.log
      : DEFAULT_FIX_OPTIONS.debug,
})
