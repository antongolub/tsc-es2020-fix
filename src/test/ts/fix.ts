import * as cp from 'child_process'
import fse from 'fs-extra'
import { globbySync } from 'globby'
import { dirname, resolve } from 'node:path'
import { fileURLToPath } from 'node:url'
import semver from 'semver'
import { temporaryDirectory } from 'tempy'

import {
  DEFAULT_FIX_OPTIONS,
  fix,
  fixBlankFiles,
  fixContents,
  fixDirnameVar,
  fixFilenameVar,
  fixModuleReferences,
} from '../../main/ts'
import { normalizeOptions } from '../../main/ts/fix'
import { read } from '../../main/ts/util'

const __dirname = dirname(fileURLToPath(import.meta.url))
const fakeProject = resolve(__dirname, '../fixtures/ts-project')
const temp = temporaryDirectory()

afterAll(() => {
  fse.removeSync(temp)
})

describe('normalizeOptions()', () => {
  it('merges DEFAULT_FIX_OPTIONS with specified opts input', () => {
    expect(normalizeOptions()).toEqual(DEFAULT_FIX_OPTIONS)
    expect(
      normalizeOptions({
        tsconfig: ['foo.json'],
        dirnameVar: false,
      }),
    ).toEqual({
      cwd: process.cwd(),
      debug: expect.any(Function),
      tsconfig: ['foo.json'],
      dirnameVar: false,
      filenameVar: true,
      ext: true,
      unlink: true,
    })
  })
})

describe('patches', () => {
  describe('fix()', () => {
    it('patches ts sources as required by opts', async () => {
      const cwd = resolve(temp, 't1')

      fse.copySync(fakeProject, cwd)

      await fix({
        cwd,
        src: ['src'],
        ext: '.js',
        debug: true,
        dirnameVar: false,
        filenameVar: false,
      })

      expect(read(resolve(temp, 't1/src/main/ts/index.ts'))).toMatchSnapshot()
    })

    it('patches target (tsc-compiled) files as required by opts', async () => {
      const cwd = resolve(temp, 'from')
      const out = resolve(temp, 'from')

      fse.copySync(fakeProject, cwd)

      await fix({
        cwd,
        out,
        tsconfig: ['tsconfig.es5.json', 'tsconfig.es6.json'],
        ext: '.mjs',
        debug: false,
      })

      const res = resolve(temp, 'from/target/es6/index.mjs')
      const contents = read(res)

      // NodeJS v16.14.0+ requires {assert: {type: 'json'}} for json imports
      // https://github.com/nodejs/node/pull/40250
      // https://github.com/nodejs/node/commit/7b996655cfcb37d732eca3f61c51d701cd97d8d1
      if (semver.gte('16.14.0', cp.execSync('node -v').toString().trim())) {
        fse.writeFileSync(res, contents.replace(` assert { type: 'json' };`, ''))
      }

      expect(contents).toMatchSnapshot()
      expect(
        cp
          .execSync(`node --experimental-top-level-await --experimental-json-modules ${res}`, {
            cwd: temp,
            env: {},
            timeout: 5000,
          })
          .toString()
          .trim(),
      ).toBe('barbaz')
    })
  })

  describe('contents', () => {
    const files = globbySync(
      [
        'target/**/*.(m|c)?js',
        'node_modules/**/*.(m|c)?js',
        '!node_modules/e2',
        '!node_modules/**/node_modules/**/*.(m|c)?js',
      ],
      {
        cwd: fakeProject,
        onlyFiles: true,
        absolute: true,
      },
    )
    const file = resolve(fakeProject, 'target/es6/index.js')
    const content = read(file)

    it('fixRelativeModuleReferences() appends file ext to module refs except for the ones that declare "exports" in pkg.json', () => {
      expect(
        fixModuleReferences(content, file, files, fakeProject),
      ).toMatchSnapshot()
    })

    it('fixDirnameVar() replaces __dirname refs', () => {
      expect(fixDirnameVar(content)).toMatchSnapshot()
    })

    it('fixFilenameVar() replaces __filename refs', () => {
      expect(fixFilenameVar(content)).toMatchSnapshot()
    })

    it('fixBlankFiles() replaces empty contents with empty export statement', () => {
      expect(fixBlankFiles('')).toMatchSnapshot()
    })

    it('fixContents() assembles all content modifiers', () => {
      expect(
        fixContents(content, file, files, {
          ...DEFAULT_FIX_OPTIONS,
          cwd: fakeProject,
        }),
      ).toMatchSnapshot()
    })

    it('fixContents() with no flags does not provide any effects', () => {
      expect(
        fixContents(content, file, files, {
          ext: false,
          cwd: fakeProject,
          filenameVar: false,
          dirnameVar: false,
          tsconfig: './tsconfig.json',
          debug: () => {}, // eslint-disable-line
        }),
      ).toEqual(content)
    })
  })
})
