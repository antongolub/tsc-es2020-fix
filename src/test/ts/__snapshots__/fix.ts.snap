// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`patches contents fixBlankFiles() replaces empty contents with empty export statement 1`] = `
"
export {}
export default undefined
"
`;

exports[`patches contents fixContents() assembles all content modifiers 1`] = `
"import { createRequire } from 'module';
const require = createRequire(import.meta.url);
import { foo } from './foo.js';
import './bar.js';
import * as e1def from 'e1/a/b/c/index.js';
import * as e1root from 'e1';
export { default as json } from './json-data.json' assert { type: 'json' };
const { e1 } = e1def;
const { e1: e1x } = e1root;
export { e1, e1x };
export { m1 } from 'm1';
export { m1 as m1x } from 'm1/index.js';
export { e2 } from 'e2';
export { e2 as es3 } from 'e2/index';
export { e2 as es4 } from 'e2/alias';
export { e2foo } from 'e2/foo';
export { e2bar } from 'e2/bar-bundle';
export * from './foo.js';
export * from './baz/index.js';
export * from './q/u/x/index.js';
export const foobaz = foo + 'baz';
export { foo as foo1 } from './foo.js';
export const foo2 = await import('./foo.js');
export { qux } from './qux.js/index.js';
export const dirname = /file:\\\\/\\\\/(.+)\\\\/[^/]/.exec(import.meta.url)[1];
export const filename = /file:\\\\/\\\\/(.+)/.exec(import.meta.url)[1];
export const rfoo = require('e1/a/b/c/index.js');
export * from './only-types.js';
console.log(foobaz);
//# sourceMappingURL=index.js.map"
`;

exports[`patches contents fixContents() replaces \`.\` with \`./index.cjs\` 1`] = `
"export * from './index.cjs';
export * as reexport from './index.cjs';
//# sourceMappingURL=index-ref.js.map"
`;

exports[`patches contents fixDefaultExport() injects export default 1`] = `
"export {};
export default undefined
"
`;

exports[`patches contents fixDirnameVar() replaces __dirname refs 1`] = `
"import { createRequire } from 'module';
const require = createRequire(import.meta.url);
import { foo } from './foo';
import './bar';
import * as e1def from 'e1/a/b/c';
import * as e1root from 'e1';
export { default as json } from './json-data.json' assert { type: 'json' };
const { e1 } = e1def;
const { e1: e1x } = e1root;
export { e1, e1x };
export { m1 } from 'm1';
export { m1 as m1x } from 'm1/index';
export { e2 } from 'e2';
export { e2 as es3 } from 'e2/index';
export { e2 as es4 } from 'e2/alias';
export { e2foo } from 'e2/foo';
export { e2bar } from 'e2/bar-bundle';
export * from './foo';
export * from './baz';
export * from './q/u/x';
export const foobaz = foo + 'baz';
export { foo as foo1 } from './foo.js';
export const foo2 = await import('./foo');
export { qux } from './qux.js';
export const dirname = /file:\\\\/\\\\/(.+)\\\\/[^/]/.exec(import.meta.url)[1];
export const filename = __filename;
export const rfoo = require('e1/a/b/c');
export * from './only-types';
console.log(foobaz);
//# sourceMappingURL=index.js.map"
`;

exports[`patches contents fixFilenameVar() replaces __filename refs 1`] = `
"import { createRequire } from 'module';
const require = createRequire(import.meta.url);
import { foo } from './foo';
import './bar';
import * as e1def from 'e1/a/b/c';
import * as e1root from 'e1';
export { default as json } from './json-data.json' assert { type: 'json' };
const { e1 } = e1def;
const { e1: e1x } = e1root;
export { e1, e1x };
export { m1 } from 'm1';
export { m1 as m1x } from 'm1/index';
export { e2 } from 'e2';
export { e2 as es3 } from 'e2/index';
export { e2 as es4 } from 'e2/alias';
export { e2foo } from 'e2/foo';
export { e2bar } from 'e2/bar-bundle';
export * from './foo';
export * from './baz';
export * from './q/u/x';
export const foobaz = foo + 'baz';
export { foo as foo1 } from './foo.js';
export const foo2 = await import('./foo');
export { qux } from './qux.js';
export const dirname = __dirname;
export const filename = /file:\\\\/\\\\/(.+)/.exec(import.meta.url)[1];
export const rfoo = require('e1/a/b/c');
export * from './only-types';
console.log(foobaz);
//# sourceMappingURL=index.js.map"
`;

exports[`patches contents fixRelativeModuleReferences() appends file ext to module refs except for the ones that declare "exports" in pkg.json 1`] = `
"import { createRequire } from 'module';
const require = createRequire(import.meta.url);
import { foo } from './foo.js';
import './bar.js';
import * as e1def from 'e1/a/b/c/index.js';
import * as e1root from 'e1';
export { default as json } from './json-data.json' assert { type: 'json' };
const { e1 } = e1def;
const { e1: e1x } = e1root;
export { e1, e1x };
export { m1 } from 'm1';
export { m1 as m1x } from 'm1/index.js';
export { e2 } from 'e2';
export { e2 as es3 } from 'e2/index';
export { e2 as es4 } from 'e2/alias';
export { e2foo } from 'e2/foo';
export { e2bar } from 'e2/bar-bundle';
export * from './foo.js';
export * from './baz/index.js';
export * from './q/u/x/index.js';
export const foobaz = foo + 'baz';
export { foo as foo1 } from './foo.js';
export const foo2 = await import('./foo.js');
export { qux } from './qux.js/index.js';
export const dirname = __dirname;
export const filename = __filename;
export const rfoo = require('e1/a/b/c/index.js');
export * from './only-types.js';
console.log(foobaz);
//# sourceMappingURL=index.js.map"
`;

exports[`patches fix() patches target (tsc-compiled) files as required by opts 1`] = `
"import './bar';
export { default as json } from './json-data.json';
declare const e1: string;
declare const e1x: string;
export { e1, e1x };
export { m1 } from 'm1';
export { m1 as m1x } from 'm1/index.js';
export { e2 } from 'e2';
export { e2 as es3 } from 'e2/index';
export { e2 as es4 } from 'e2/alias';
export { e2foo } from 'e2/foo';
export { e2bar } from 'e2/bar-bundle';
export * from './foo.js';
export * from './baz/index.js';
export * from './q/u/x/index.js';
export declare const foobaz: string;
export { foo as foo1 } from './foo.js';
export declare const foo2: typeof import(\\"./foo.js\\");
export { qux } from './qux.js/index.js';
export declare const dirname: string;
export declare const filename: string;
export declare const rfoo: any;
export * from './only-types.js';
"
`;

exports[`patches fix() patches target (tsc-compiled) files as required by opts 2`] = `
"export {};
//# sourceMappingURL=only-types.mjs.map"
`;

exports[`patches fix() patches target (tsc-compiled) files as required by opts 3`] = `
"import { createRequire } from 'module';
const require = createRequire(import.meta.url);
import { foo } from './foo.mjs';
import './bar.mjs';
import * as e1def from 'e1/a/b/c/index.js';
import * as e1root from 'e1';
export { default as json } from './json-data.json' assert { type: 'json' };
const { e1 } = e1def;
const { e1: e1x } = e1root;
export { e1, e1x };
export { m1 } from 'm1';
export { m1 as m1x } from 'm1/index.js';
export { e2 } from 'e2';
export { e2 as es3 } from 'e2/index';
export { e2 as es4 } from 'e2/alias';
export { e2foo } from 'e2/foo';
export { e2bar } from 'e2/bar-bundle';
export * from './foo.mjs';
export * from './baz/index.mjs';
export * from './q/u/x/index.mjs';
export const foobaz = foo + 'baz';
export { foo as foo1 } from './foo.mjs';
export const foo2 = await import('./foo.mjs');
export { qux } from './qux.js/index.mjs';
export const dirname = /file:\\\\/\\\\/(.+)\\\\/[^/]/.exec(import.meta.url)[1];
export const filename = /file:\\\\/\\\\/(.+)/.exec(import.meta.url)[1];
export const rfoo = require('e1/a/b/c/index.js');
export * from './only-types.mjs';
console.log(foobaz);
//# sourceMappingURL=index.mjs.map"
`;

exports[`patches fix() patches ts sources as required by opts 1`] = `
"import { createRequire } from 'module'
const require = createRequire(import.meta.url)

import { foo } from './foo.js'

import './bar.js'

import * as e1def from 'e1/a/b/c/index.js'

import * as e1root from 'e1'

export { default as json } from './json-data.json' assert {type: 'json'}

const { e1 } = e1def

const { e1: e1x } = e1root

export { e1, e1x }

export { m1 } from 'm1'

export { m1 as m1x } from 'm1/index.js'

export { e2 } from 'e2'

export { e2 as es3 } from 'e2/index'

export { e2 as es4 } from 'e2/alias'

export { e2foo } from 'e2/foo'

export { e2bar } from 'e2/bar-bundle'

export * from './foo.js'

export * from './baz/index.js'

export * from './q/u/x/index.js'

export const foobaz = foo + 'baz'

export { foo as foo1 } from './foo.js'

export const foo2 = await import('./foo.js')

export { qux } from './qux.js/index.js'

export const dirname = __dirname

export const filename = __filename

export const rfoo = require('e1/a/b/c/index.js')

export * from './only-types.js'

console.log(foobaz)
"
`;

exports[`patches fix() patches ts sources as required by opts 2`] = `
"export * from './index.js'
export * as reexport from './index.js'
"
`;

exports[`patches fix() patches ts sources as required by opts 3`] = `"export * from '../index.js'"`;
